src/Features/Core/Portable/Diagnostics/Analyzers/QualifyMemberAccessDiagnosticAnalyzerBase.cs

```diff
@@ -0,0 +1,110 @@
+        private static readonly DiagnosticDescriptor s_descriptorQualifyMemberAccess = new DiagnosticDescriptor(IDEDiagnosticIds.AddQualificationDiagnosticId,
+                                                                    s_localizableTitleQualifyMembers,
+                                                                    s_localizableMessage,
+                                                                    DiagnosticCategory.Style,
+                                                                    DiagnosticSeverity.Hidden,
+                                                                    isEnabledByDefault: true,
+                                                                    customTags: DiagnosticCustomTags.Unnecessary);
+
+        private OptionSet _lazyDefaultOptionSet;
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(s_descriptorQualifyMemberAccess);
+
+        protected abstract ImmutableArray<TLanguageKindEnum> GetSupportedSyntaxKinds();
+        protected abstract bool IsAlreadyQualifiedMemberAccess(SyntaxNode node);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterOperationAction(AnalyzeOperation, OperationKind.FieldReferenceExpression, OperationKind.PropertyReferenceExpression, OperationKind.MethodReferenceExpression);
+        }
+
+        private void AnalyzeOperation(OperationAnalysisContext context)
+        {
+            var memberReference = (IMemberReferenceExpression)context.Operation;
+            if (memberReference.Instance == null)
+            {
+                return;
+            }
+
+            if (IsAlreadyQualifiedMemberAccess(memberReference.Instance.Syntax))
+            {
+                return;
+            }
+
+            var optionSet = GetOptionSet(context.Options);
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var memberSymbol = GetSymbol(memberReference);
+            if (memberSymbol == null)
+            {
+                return;
+            }
+
+            if (memberSymbol.ContainingSymbol != memberReference.Instance.ResultType)
+            {
+                return;
+            }
+
+            var language = context.Operation.Syntax.Language;
+            if (!SimplificationHelpers.ShouldSimplifyMemberAccessExpression(memberSymbol, language, optionSet))
+            {
+                context.ReportDiagnostic(Diagnostic.Create(s_descriptorQualifyMemberAccess, context.Operation.Syntax.GetLocation()));
+            }
+        }
+
+        private static ISymbol GetSymbol(IMemberReferenceExpression memberReferenceExpression)
+        {
+            switch (memberReferenceExpression.Kind)
```

Comment:
Put a comment in that this can be simplified once IMemberReferenceExpression just exposes 'Member'